xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,148 @@
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import arith, func, scf
+from xdsl.dialects.builtin import IndexType, ModuleOp, f32, i32
+from xdsl.dialects.experimental import math
+from xdsl.dialects.printf import PrintCharOp, PrintIntOp
+from xdsl.ir import Block, MLContext
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.printer import Printer
+
+
+class LowerPrintCharToPutchar(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_char to an
+    external function call
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintCharOp, rewriter: PatternRewriter, /):
+        func_call = func.Call.get("putchar", [op.char], [i32])
+        # Add empty new_results, since a result is necessary for linking
+        # putchar, but the result does not exist anywhere.
+        rewriter.replace_matched_op(func_call, new_results=[])
+
+
+class ConvertPrintIntToItoa(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_int to an
+    inline_itoa function
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintIntOp, rewriter: PatternRewriter, /):
+        func_call = func.Call.get("inline_itoa", [op.int], [])
+        rewriter.replace_matched_op(func_call)
+
+
+def get_inline_itoa():
```

Comment:
I can't think of any other way to make this "prettier" other than trying to factor out parts of it to separate utility methods.
This might make testing easier as well, building incrementally.